File: sdk/python/tempmail_sdk/html_utils.py
"""
HTML text extraction helpers used by providers and normalization.
"""

from html.parser import HTMLParser


class _HTMLTextExtractor(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self._parts: list[str] = []
        self._skip_depth = 0

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        del attrs
        name = tag.lower()
        if name in {"script", "style"}:
            self._skip_depth += 1
            return
        self._parts.append(" ")

    def handle_endtag(self, tag: str) -> None:
        name = tag.lower()
        if name in {"script", "style"} and self._skip_depth:
            self._skip_depth -= 1
            return
        self._parts.append(" ")

    def handle_data(self, data: str) -> None:
        if self._skip_depth:
            return
        self._parts.append(data)

    def text(self) -> str:
        return " ".join("".join(self._parts).split())


def html_to_text(value: str) -> str:
    parser = _HTMLTextExtractor()
    parser.feed(value or "")
    parser.close()
    return parser.text()

File: sdk/ruby/lib/tempmail_sdk/html_utils.rb
# frozen_string_literal: true

module TempmailSdk
  # HTML 文本提取辅助
  module HtmlUtils
    module_function

    # 将 HTML 粗略转换为纯文本：剥离 script/style，去标签，压缩空白
    # @param value [String, nil]
    # @return [String]
    def html_to_text(value)
      return "" if value.nil? || value.empty?

      s = value.gsub(%r{<script\b[^>]*>.*?</script>}mi, " ")
      s = s.gsub(%r{<style\b[^>]*>.*?</style>}mi, " ")
      s = s.gsub(/<[^>]+>/, " ")
      s = unescape_entities(s)
      s.split.join(" ")
    end

    # 转义 HTML 特殊字符
    # @param text [String]
    # @return [String]
    def escape(text)
      text.to_s
          .gsub("&", "&amp;")
          .gsub("<", "&lt;")
          .gsub(">", "&gt;")
          .gsub('"', "&quot;")
    end

    # 反转义常见 HTML 实体
    def unescape_entities(text)
      text.gsub("&amp;", "&")
          .gsub("&lt;", "<")
          .gsub("&gt;", ">")
          .gsub("&quot;", '"')
          .gsub("&#39;", "'")
          .gsub("&nbsp;", " ")
    end
  end
end
